### PR TITLE
feat: support workflow 

### PR DESCRIPTION
**Pull request by alexcodelf**
**Problem:**
Support workflow in walrus.

**Solution:**
This PR is going to 
1. refactor terraform deployer, add service revision manager
2. introduce workflow management of walrus that based on argo workflow framework
  - approval workflow step
  - service workflow step
  - support workflow orchestration
  - add step level timeout and retry strategy

**Related Issue:**
#734

### CODE REVIEW

File: pack/server/image/Dockerfile [hunk 1]
```diff
@@ -152,11 +152,12 @@ COPY --from=sealio/casdoor:v1.344.0-seal.1 \
     /casdoor \
     /usr/bin/
 
-# finops tools
+# finops and workflows tools
```
Comment (gitlawr) [inline pack/server/image/Dockerfile:155]: set them apart

File: pkg/apis/service/basic.go [hunk 2]
```diff
@@ -31,13 +31,12 @@ func (h Handler) Create(req CreateRequest) (CreateResponse, error) {
 	}
 
 	createOpts := pkgservice.Options{
-		TlsCertified: h.tlsCertified,
+		ModelClient: h.modelClient,
```
Comment (gitlawr) [inline pkg/apis/service/basic.go:34]: I prefer not putting them in opts given that we explicitly pass modelClient in most of the code base.

File: pkg/environment/helper.go [hunk 3]
```diff
@@ -22,3 +26,28 @@ func GetManagedNamespaceName(e *model.Environment) string {
 
 	return fmt.Sprintf("%s-%s", e.Edges.Project.Name, e.Name)
 }
+
+func GetConnectors(ctx context.Context, mc model.ClientSet, environmentID object.ID) (model.Connectors, error) {
```
Comment (gitlawr) [inline pkg/environment/helper.go:30]: Can we put it to dao/environment.go

File: pkg/service/dependency.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+	"github.com/seal-io/walrus/pkg/dao/model/service"
+	"github.com/seal-io/walrus/pkg/dao/model/servicerelationship"
+	"github.com/seal-io/walrus/pkg/dao/model/servicerevision"
+	"github.com/seal-io/walrus/pkg/dao/types"
+	"github.com/seal-io/walrus/pkg/dao/types/object"
+	"github.com/seal-io/walrus/pkg/terraform/parser"
+)
+
+// GetServiceDependencyOutputsByID gets the dependency outputs of the service by service id.
+func GetServiceDependencyOutputsByID(
```
Comment (gitlawr) [inline pkg/service/dependency.go:19]: Can we put it to dao/service.go

File: pkg/servicerevision/plan.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
+package servicerevision
```
Comment (gitlawr) [inline pkg/servicerevision/plan.go:1]: Can you explain purporse of the `refactor: deployer and modify revision status` commit? Better to know in commit message

File: pkg/dao/types/built_in_names.go [hunk 6]
```diff
@@ -4,3 +4,8 @@ const (
 	WalrusSystemNamespace      = "walrus-system"
 	DeployerServiceAccountName = "walrus-deployer"
 )
+
+const (
+	WalrusWorkflowNamespace    = "walrus-workflow"
```
Comment (gitlawr) [inline pkg/dao/types/built_in_names.go:9]: Can we use WalrusSystemNamespace, is there permission concern?

File: pkg/apis/workflow/extension.go [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+	}
+
+	if wf == nil {
+		return nil, nil
+	}
+
+	return model.ExposeWorkflowExecution(wf), nil
+}
+
+func (h Handler) RouteApplyRequest(req RouteApplyRequest) (RouteApplyResponse, error) {
```
Comment (gitlawr) [inline pkg/apis/workflow/extension.go:39]: Does Apply mean run the workflow? Can we call it RouteRun?

File: pkg/apis/workflowexecution/extension_view.go [hunk 8]
```diff
@@ -0,0 +1,13 @@
+package workflowexecution
+
+import "github.com/seal-io/walrus/pkg/dao/model"
+
+type RouteResubmitRequest struct {
+	_ struct{} `route:"PUT=/resubmit"`
```
Comment (gitlawr) [inline pkg/apis/workflowexecution/extension_view.go:6]: Can we call it rerun?

File: pkg/apis/service/extension.go [hunk 9]
```diff
@@ -472,3 +477,175 @@ func (h Handler) RouteGetGraph(req RouteGetGraphRequest) (*RouteGetGraphResponse
 		Edges:    edges,
 	}, nil
 }
+
+// CollectionRouteWorkflowStepExecutionRequest is the request for workflow service step execution.
+// It will create or update the service and return the revision config file.
+func (h Handler) CollectionRouteWorkflowStepExecution(req CollectionRouteWorkflowStepExecutionRequest) (any, error) {
```
Comment (gitlawr) [inline pkg/apis/service/extension.go:483]: Why service API needs to be aware of workflow execution? I was expecting workflow itself handles deployment using service CRUD APIs.

File: pkg/service/dependency.go (hunk 4 above)
Comment (alexcodelf) [reply pkg/service/dependency.go:19]: This function depends on terraform/parser， put into dao will cause import cycle

File: pkg/apis/service/basic.go (hunk 2 above)
Comment (alexcodelf) [reply pkg/apis/service/basic.go]: updated

File: pack/server/image/Dockerfile (hunk 1 above)
Comment (alexcodelf) [reply pack/server/image/Dockerfile]: updated

File: pkg/dao/types/status/status_servicerevision.go [hunk 10]
```diff
@@ -1,38 +1,31 @@
...
 
 const (
-	ServiceRevisionStatusReady ConditionType = "Ready"
+	ServiceRevisionStatusRunning ConditionType = "Running"
+	ServiceRevisionStatusReady   ConditionType = "Ready"
 
 	ServiceRevisionSummaryStatusRunning string = "Running"
 	ServiceRevisionSummaryStatusFailed  string = "Failed"
-	ServiceRevisionSummaryStatusSucceed string = "Succeed"
+	ServiceRevisionSummaryStatusSucceed string = "Ready"
```
Comment (gitlawr) [inline pkg/dao/types/status/status_servicerevision.go:9]: Can you explain the condition changes?

File: pkg/dao/schema/workflow_step.go [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+			NotEmpty().
+			Immutable(),
+		object.IDField("workflow_id").
+			Comment("ID of the workflow that this workflow step belongs to.").
+			NotEmpty().
+			Immutable(),
+		object.IDField("workflow_stage_id").
+			Comment("ID of the stage that this workflow step belongs to.").
+			Immutable(),
+		field.JSON("spec", map[string]any{}).
```
Comment (gitlawr) [inline pkg/dao/schema/workflow_step.go:42]: Should we call it attributes to align with other APIs?

File: pkg/dao/schema/workflow_step.go [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+			Comment("Spec of the workflow step.").
+			Optional(),
+		field.JSON("input", map[string]any{}).
```
Comment (gitlawr) [inline pkg/dao/schema/workflow_step.go:45]: inputs

File: pkg/dao/schema/workflow_step.go [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+			Comment("Input of the workflow step.").
+			Optional(),
+		field.JSON("output", map[string]any{}).
```
Comment (gitlawr) [inline pkg/dao/schema/workflow_step.go:48]: outputs

File: pkg/dao/schema/workflow_step_execution.go [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+			Default(1).
+			Annotations(
+				entx.SkipIO()),
+		field.Time("execute_time").
+			Comment("Time of the step execution started.").
+			Optional().
+			Annotations(
+				entx.SkipInput()),
+		field.Int("duration").
+			Comment("Duration of the workflow step execution.").
```
Comment (gitlawr) [inline pkg/dao/schema/workflow_step_execution.go:64]: in second? make it clear in comment

File: pkg/workflow/step/types/new.go
Comment (gitlawr) [inline pkg/workflow/step/types/new.go:1]: find a better file name

Comment (alexcodelf) [reply pkg/workflow/step/types/new.go:1]: There are other packages have the file call this name, shall we rename them together?

File: pkg/dao/types/status/status_servicerevision.go (hunk 10 above)
Comment (alexcodelf) [reply pkg/dao/types/status/status_servicerevision.go:9]: refactor revision and deployer will not influence workflow feature, I will create another Pull Request.

File: pkg/workflow/step/types/new.go
Comment (gitlawr) [reply pkg/workflow/step/types/new.go:1]: keep it then.

**Review by gitlawr — APPROVED** [review]
LGTM

File: pkg/dao/schema/workflow.go [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+			Positive().
+			Default(10),
+		field.Int("timeout").
+			Comment("Timeout seconds of the workflow.").
+			NonNegative().
+			Default(0),
+		field.Int("version").
+			Comment("Execution version of the workflow.").
+			NonNegative().
+			Default(0).
```
Comment (aiwantaozi) [inline pkg/dao/schema/workflow.go:58]: what does the default version 0 mean? `WorkflowExecution` doesn't have a default config?

File: pkg/dao/types/status/status_workflow_execution.go [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+	WorkflowExecutionStatusPending ConditionType = "Pending"
+	WorkflowExecutionStatusRunning ConditionType = "Running"
+)
+
+// workflowExecutionStatusPaths makes the following decision.
+//
+//	|  Condition Type  |     Condition Status    | Human Readable Status | Human Sensible Status |
+//	| ---------------- | ----------------------- | --------------------- | --------------------- |
+//	| Pending          | Unknown                 | Pending               | Transitioning         |
+//	| Pending          | False                   | PendingFailed         | Error                 |
```
Comment (aiwantaozi) [inline pkg/dao/types/status/status_workflow_execution.go:13]: PendingFailed seems strange, PendingFailed -> Failed?

File: pkg/dao/types/status/status_workflow_execution.go [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+	WorkflowStageExecutionStatusPending ConditionType = "Pending"
+	WorkflowStageExecutionStatusRunning ConditionType = "Running"
+)
+
+// workflowStageExecutionStatusPaths makes the following decision.
+//
+//	|  Condition Type  |     Condition Status    | Human Readable Status | Human Sensible Status |
+//	| ---------------- | ----------------------- | --------------------- | --------------------- |
+//	| Pending          | Unknown                 | Pending               | Transitioning         |
+//	| Pending          | False                   | PendingFailed         | Error                 |
```
Comment (aiwantaozi) [inline pkg/dao/types/status/status_workflow_execution.go:41]: ditto

File: pkg/dao/types/status/walker.go [hunk 18]
```diff
@@ -215,6 +215,8 @@ func getGeneralDecide[T ~string](step T) Decide {
 var replacements = map[string]struct {
 	T, E, D string
 }{
+	"Running":     {"Running", "Failed", "Completed"},
+	"Pending":     {"Pending", "Failed", "Pended"},
```
Comment (aiwantaozi) [inline pkg/dao/types/status/walker.go:219]: What status will display while the Pending status is True? Pended -> Pending, please double check here?

File: pkg/workflow/step/service_step.go [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+set -o pipefail
+
+serverURL="{{workflow.parameters.server}}"
+projectID="{{workflow.parameters.projectID}}"
+environmentID="{{inputs.parameters.environmentID}}"
+token="{{inputs.parameters.token}}"
+jobType="{{inputs.parameters.jobType}}"
+attributes='{{inputs.parameters.attributes}}'
+serviceName="{{inputs.parameters.serviceName}}"
+commomPath="$serverURL/v1/projects/$projectID/environments/$environmentID"
```
Comment (aiwantaozi) [inline pkg/workflow/step/service_step.go:32]: commonPath?

File: pkg/dao/schema/workflow.go [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+			NotEmpty().
+			Immutable(),
+		field.Int("parallelism").
+			Comment("Number of task pods that can be executed in parallel of workflow.").
+			Positive().
+			Default(10),
+		field.Int("timeout").
+			Comment("Timeout seconds of the workflow.").
+			NonNegative().
+			Default(0),
```
[the remark this answers is not in the document]
Comment (alexcodelf) [reply pkg/dao/schema/workflow.go:54]: 0 means there is no limit of timeout. Currently, Front end will configure 30min, 1hour, 2hour .. etc as select options for users to select if they want.

File: pkg/apis/workflow/extension.go [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Update workflow version.
+	wf, err = h.modelClient.Workflows().UpdateOne(wf).
+		AddVersion(1).
```
Comment (aiwantaozi) [inline pkg/apis/workflow/extension.go:73]: Do you think we should add row level lock here?
https://entgo.io/docs/feature-flags/#row-level-locks

File: pkg/dao/schema/workflow.go (hunk 15 above)
Comment (alexcodelf) [reply pkg/dao/schema/workflow.go:58]: Version means the current execution sequence. When a workflow is created, there is no workflow execution. Therefore the version of the workflow is zero. When workflow execution created, the version will add 1.

**Review by aiwantaozi — APPROVED** [review]

**Pull request merged**